### PR TITLE
Extend Record Merging to multiple records

### PR DESCRIPTION
**Pull request by acwhite211**
Extend Record Merging to multiple records to allow for the handling of businesses rules and keeping the whole merging process inside of one ACID transaction.

### CODE REVIEW

**Comment by acwhite211** [comment]
Still working on creating more unit tests, but please go ahead and start testing with front-end integration. 
The old url `/api/specify/old_model_id/new_model_id/` still exists, the new extended record merging url is `/api/specify/new_model_id/` with a json body containing old_model_ids and new_record_data

**Comment by acwhite211** [comment]
@melton-jason feel free to add more unit tests

**Comment by specifysoftware** [comment]
This pull request has been mentioned on **Specify Community Forum**. There might be relevant details there:

https://discourse.specifysoftware.org/t/record-merging-in-specify-7/939/11

File: specifyweb/specify/views.py
Comment (maxxxxxdlp) [inline specifyweb/specify/views.py:517]: are type annotations required by `mypy` even when can be trivially inferred?

File: specifyweb/specify/views.py
Comment (maxxxxxdlp) [inline specifyweb/specify/views.py:640]: there should be an easier way to do this.
Something like this:
```py
if 'version' in request.GET:
  version = request.GET['version']
else:
  version = getattr(spmodels, model_name.title()).objects.get(id=new_model_id).version
```

or maybe even this, to avoid needless if/else:
```python
default_version = getattr(spmodels, model_name.title()).objects.get(id=new_model_id).version
version = request.GET.get('version', default_version)
```

File: specifyweb/specify/views.py
Comment (maxxxxxdlp) [inline specifyweb/specify/views.py:527]: should permissions be checked before `.delete()` and `.put_resource()` were called?

File: specifyweb/specify/views.py
Comment (maxxxxxdlp) [inline specifyweb/specify/views.py:636]: I am not too familiar with django/python/sqlalcheme, but it this code being run as a transaction? i.e., if some records get deleted, but then it falls mid-way, (potentially because of missing permissions), do those records get un-deleted?

additionally, if those records have attachments, does the back-end send requests to the asset server to delete the attachments? (it might have triggers doing that) that might cause issues if request doesn't go though fully

**Review by maxxxxxdlp — CHANGES_REQUESTED** [review]

File: specifyweb/specify/urls.py
Comment (maxxxxxdlp) [reply specifyweb/specify/urls.py:12]: having a UI that makes requests would also make it easier for you to construct test cases

File: specifyweb/specify/views.py
Comment (acwhite211) [reply specifyweb/specify/views.py:636]: ya, I made sure to research it.  The `@transaction.atomic` decorator encapsulates a database actions into one ACID transactions.  Even in the case of recursively calling the function, creating nested transactions, either all the actions occur, or none of them.  I'm not sure about the attachments, it's not something I can unit test easily.  We should make sure to test that testcase in the integration testing.  I'll look to see if I can find any triggers for deleting attachments, but I don't think it would cause a trigger to occur until the full atomic transaction is completed.

File: specifyweb/specify/views.py
Comment (melton-jason) [inline specifyweb/specify/views.py:533]: Looks like there's a typo here (missing an e in 'Replace')

File: specifyweb/specify/views.py
Comment (melton-jason) [inline specifyweb/specify/views.py:579]: What is the purpose of this `request_params` object? It's not used anywhere else in the function.

File: specifyweb/specify/api_tests.py
Comment (melton-jason) [inline specifyweb/specify/api_tests.py:741]: I'm not terribly familiar with this, but does Django properly convert this to a Date Time object, or leave it as a string? 
It may matter when we make Django time-aware, and it's good to have the data as closely representative to what it would be in the 'real-world' application

**Review by melton-jason — CHANGES_REQUESTED** [review]
I will need to take a more indepth look at this and add a unit test for when Agents have Group Person records. 
For a quick thrice-over, everything looks good, but I do have some questions.

**Review by grantfitzsimmons — CHANGES_REQUESTED** [review]
image
https://kufish51623-agent-merging-extremos.test.specifysystems.org/specify/overlay/merge/Agent/?records=3%2C5%2C6%2C7%2C8

After waiting for a few minutes, it seems that the merging process failed. It showed a red bar at the top of the dialog, but no error was reported to the user.

When opening the console, it seems there were some errors reported there.

Console Log:
[kufish51623-agent-merging-extremos.test.specifysystems.org-16842.zip](https://github.com/specify/specify7/files/11490881/kufish51623-agent-merging-extremos.test.specifysystems.org-16842.zip)

**Comment by maxxxxxdlp** [comment]
@melton-jason
Three cases:
- None of the agents are connected to any users - no issue
- One of the agents is connected to a user - link the user to the merged agent instead
- More than one agent is assigned to a (same or different) user - quite unlikely. bail out with a nice error message

**Comment by melton-jason** [comment]
Yes that would be a nice way of going about this particular case. 
Some more context on the `More than one agent is assigned to a (same or different) user` case; I think having an error message in this case would be fine. 
This operation is not supported in the UI at all (I do not know if it is supported in Specify 6 or not): you can not assign an agent to more than one user. 

We will need to think of ways to generalize cases such as these (if broad generalization can exist to cover _all_ backend deletion Business Rules). 
Here are some options we can take a look into:
- Utilize Django Signals
- Extend Deletion Rules
- Behavior based on attributes in the data model

**Comment by maxxxxxdlp** [comment]
> This operation is not supported in the UI at all (I do not know if it is supported in Specify 6 or not): you can not assign an agent to more than one user.

I'm afraid you misunderstood my comment. Indeed, you can't assign the same agent to multiple users.
But, the agent merging tool deals with merging multiple agents into one, thus it's entirely possible (though hopefully not likely) that you would try to merge two agents who both are assigned to a user

**Comment by maxxxxxdlp** [comment]
@melton-jason the simplest solution for this issue:
The front-end for agent merging has code for picking which agent should be considered "main"
We should just change it to consider agent connected to a user to be the main one
See the (untested) fix I pushed

**Comment by melton-jason** [comment]
> @melton-jason the simplest solution for this issue: The front-end for agent merging has code for picking which agent should be considered "main" We should just change it to consider agent connected to a user to be the main one See the (untested) fix I pushed

https://github.com/specify/specify7/assets/64045831/26ff18e4-07e5-4644-81a7-e2f7931161de

@maxxxxxdlp 

Unfortunately, this is not a complete fix. 
What about the case when more than one Agents are associated to separate Specify Users? 
Regardless of what it chooses as the 'main' Agent, there will always be another Agent associated with a Specifyuser that the backend will try and delete.

**Comment by melton-jason** [comment]
However, this falls under the third `More than one agent is assigned to a (same or different) user - quite unlikely. bail out with a nice error message` case. 
How likely is this in a production database?

**Comment by maxxxxxdlp** [comment]
> What about the case when more than one Agents are associated to separate Specify Users?

I don't think this is very likely
@grantfitzsimmons what do you think?
And even if this does occur, a business rule error would occur, which the back-end business error handler code has to handle anyway

**Comment by grantfitzsimmons** [comment]
> > What about the case when more than one Agents are associated to separate Specify Users?
> 
> I don't think this is very likely @grantfitzsimmons what do you think?

It is not very likely at all

**Comment by melton-jason** [comment]
> image https://kufish51623-agent-merging-extremos.test.specifysystems.org/specify/overlay/merge/Agent/?records=3%2C5%2C6%2C7%2C8
> After waiting for a few minutes, it seems that the merging process failed. It showed a red bar at the top of the dialog, but no error was reported to the user.
> 
> When opening the console, it seems there were some errors reported there.
> 
> Console Log: [kufish51623-agent-merging-extremos.test.specifysystems.org-16842.zip](https://github.com/specify/specify7/files/11490881/kufish51623-agent-merging-extremos.test.specifysystems.org-16842.zip)
> 
> image

This is still concerning, and I have been unable to identify the cause of the issue thus far. 
Besides the issue I pointed out regarding Specify Users, I have been able to merge Agents without incidence, both on a local instance and for simple Agents (which I always created beforehand) on the `kufish_5_16_23` database.

Perhaps the _most_ concerning and perplexing symptom of this issue is how _little_ it gives in terms of information to debug the problem. The only information that can be gleaned directly from the issue is the 405 (Method Not Allowed) response.  
Usually this might be caused by a misconfigured url or the wrong request type, but everything seems correct as far as these are concerned. 

@maxxxxxdlp 
Did I make a mistake in structuring the request in 03f11a1 and subsequentially 8d5f8f6?

**Comment by maxxxxxdlp** [comment]
Code looks ok. This could be caused by test panel restarting. That would explain the 421 error
405 would happen in that case because while test panel is restarting, the requests were sent to the nginx container or to test panel container, which doesn't accept POST

**Comment by maxxxxxdlp** [comment]
https://github.com/specify/specify7/pull/3474#issuecomment-1551330763 - I really don't like how the error message is formatted in the video

If I remember correctly, this is only happening when in development. And we still don't know why that's happening?

**Comment by melton-jason** [comment]
> [#3474 (comment)](https://github.com/specify/specify7/pull/3474#issuecomment-1551330763) - I really don't like how the error message is formatted in the video
> 
> If I remember correctly, this is only happening when in development. And we still don't know why that's happening?

In my free time in the past two days I have been looking into why this formatting is happening (#3343). I will update the issue shortly with what I have found in the interim.
If getting back to the correct Django error messages for development is a priority, I can focus on doing so. Or at the very least, I can devote more time to doing so.

**Comment by maxxxxxdlp** [comment]
You have a ton of other important things!
As long as it's better in production, it's not too bad

**Comment by melton-jason** [comment]
> Code looks ok. This could be caused by test panel restarting. That would explain the 421 error 405 would happen in that case because while test panel is restarting, the requests were sent to the nginx container or to test panel container, which doesn't accept POST

Yes, I first thought it might be an issue with the test panel restarting. However, I have attempted to perform that exact merge multiple times, and every time encountered the same behavior. 
It would make sense if the requests were being to NGINX or Test Panel containers, where POST requests are not allowed. If this is the case, then how do we manage other POST requests like creating records? The chances of the test panel being down _every_ time @grantfitzsimmons and I have attempted the merge is unlikely. Moreover, I did not encounter any `502 Gateway Time Out` or `TypeError: Failed to Fetch` errors during these attempted merges.

I have successfully merged a variety of Agents both locally and through the same `kufish_5_16_23` database, so the endpoint _does_ work correctly. 
I am just baffled by the fact that the error encountered by @grantfitzsimmons in https://github.com/specify/specify7/pull/3474#pullrequestreview-1429171474 provides very little helpful information. The 405 error is the _only_ information the application provides. There is no Business Rule, Django, or SQLAlchemy error. 
Trying to uncover why _that_ merge went wrong would be like trying to find a needle in _every_ haystack. 

There is a _lot_ that could go wrong, and we don't have a lot to go off of.

**Comment by melton-jason** [comment]
I have identified a case on a local copy of  `umich-mollusks` which exhibited the same behavior as in  https://github.com/specify/specify7/pull/3474#pullrequestreview-1429171474. I discovered the issue, and was able to apply a fix. 

Here is a video showing two agents which I tried to merge which showed the aforementioned behavior. 

https://github.com/specify/specify7/assets/64045831/04c67bd8-96f6-43b3-9320-1cae2ed7aefe

Assuming the error was likely due to the references, I investigated the 13 Collecting Events which reference `ex Rush`. 
In the Collecting Events, there were multiple instances where `R.C. Rush` was a Collector alongside `ex Rush`. 
I removed `R.C. Rush` from being a Collector in every Collecting Event which also referenced `ex Rush`. 

Then I attempted the merge again, and it worked! 🥳 

https://github.com/specify/specify7/assets/64045831/1ea9d9ff-fba8-4b66-acc0-e042f15dd57d

Our phantom issue is what appears to be an Integrity Error. 
I will need to investigate how the endpoint works currently, and I can work on a fix.

**Comment by melton-jason** [comment]
@maxxxxxdlp 
Do we want to merge this back to `agent-merging` before conflicts get out of hand? 😅 
You could argue that this change is within the scope of the pull request, but this pull request is already getting a tad too long.
It has accomplished what it sought out to accomplish: modify the endpoint so agent merging is only a single request, and the needed data (the JSON of the record to save, and an Array of id's from the clones')  is contained within the POST body.

**Comment by maxxxxxdlp** [comment]
yes please, let's do that

**Comment by melton-jason** [comment]
Ok. I'd say we can merge this back to `agent-merging` once @acwhite211 has gone through the review comments

**Comment by acwhite211** [comment]
Let's go ahead now and merge this with the `agent-merging` branch and change the front-end code referencing the old api with the new record-merging api so that we can start testing again.

**Review by CarolineDenis — APPROVED** [review]

File: specifyweb/frontend/js_src/lib/components/Merging/index.tsx
Comment (maxxxxxdlp) [inline specifyweb/frontend/js_src/lib/components/Merging/index.tsx:221]: This comment should no longer be here as we are no longer making requests sequentially

File: specifyweb/specify/views.py
Comment (maxxxxxdlp) [inline specifyweb/specify/views.py:517]: can be simplified to `has_new_record_info = new_record_info is not None`

**Review by maxxxxxdlp — APPROVED** [review]
Looks good. Two small refactoring things - can be resolved after merging this

File: specifyweb/frontend/js_src/lib/components/Merging/index.tsx
Comment (melton-jason) [reply specifyweb/frontend/js_src/lib/components/Merging/index.tsx]: Fixed in 2b10695

File: specifyweb/specify/views.py
Comment (melton-jason) [reply specifyweb/specify/views.py]: Fixed in 2b10695

**Comment by melton-jason** [comment]
> Perhaps the most concerning and perplexing symptom of this issue is how little it gives in terms of information to debug the problem. The only information that can be gleaned directly from the issue is the 405 (Method Not Allowed) response.
> Usually this might be caused by a misconfigured url or the wrong request type, but everything seems correct as far as these are concerned.

https://github.com/specify/specify7/pull/3474#issuecomment-1551534512

This was caused because the backend was returning an `http.HttpResponseNotAllowed()`

```python
# Determine which of the records will be assigned as old and new with the timestampcreated field
foreign_record_lst = foreign_model.objects.filter(**{field_name_id: new_model_id})
if foreign_record_lst.count() > 1:
    # NOTE: Maybe try handling multiple possible row that are potentially causes the conflict.
    # Would have to go through all constraints and check records based on columns in each constraint. 
    return http.HttpResponseNotAllowed('Error! Multiple records violating uniqueness constraints in ' + table_name) 
```
    
image

Raising an error instead of returning the HttpResponseNotAllowed would make the frontend and toasts behave normally

**Comment by maxxxxxdlp** [comment]
@melton-jason I don't think back-end is at fault here. It's front-end's fault for not handling the error correctly.

There is code that tries to figure out if the error was generated by a permission system (thus display the permission error dialog) or by session time out (thus, prompt the user to sign in instead). It's concerning that this error didn't trigger either error dialog.

Relevant code:

https://github.com/specify/specify7/blob/bd2ba160a43bfb555757a8a0cc7c82095ac85e0e/specifyweb/frontend/js_src/lib/components/Errors/FormatError.tsx#L153-L177

Can you please take a look at what's going on there?

**Pull request merged**
